### ISSUE TITLE
TypeXXX missing in a lot of CMIP6 FX files

### ISSUE THREAD

**Issue by valeriupredoi** [ESMValGroup/ESMValCore#329]
An example reported by @bascrezee in #328 and also reported by @ValerioLembo and explained by @schlunma in https://github.com/ESMValGroup/ESMValTool/pull/1272#issuecomment-544873027

Question to be asked: what is this `typeXXX` stuff? @zklaus or @jvegasbsc you guys know? If not I'll take it up with CMIP6 data gurus like Ruth Petrie at CEDA or @agstephens 

Solution at hand: write fixes seems a bit of overkill, but if this type crap is needed then we should automate it for all models; if it's not, forget about it in the CMOR checks :beer:

**Comment by schlunma** [comment]
This morning @debe-kevin and me also discovered several datasets where scalar coordinates (e.g. `depth`, `height` or the `typeXXX` ones) are missing. It would be really nice to have automatic fixes for this, because I think that a lot of models are affected.

**Comment by valeriupredoi** [comment]
one important thing we have to bear in mind is that CMIP6 data is far from final (and the quality of it differs from one ESGF node to another), so before we start writing a high complexity function that addresses automated fixes we have to think twice. But yeah, am in favour of automatic fixes -> @jvegasbsc

**Comment by zklaus** [comment]
These things (`typeXXX`) are generally related to vegetation and this is imho the reason why they are a little bit less mature than other parts of the esms that have been around for longer.

One problem that was encountered in the specification of CMIP6 was that there were standard names in CF for some of the needed categories, but not for all. This let to the slightly inhomogeneous situation were some of that information is purely conveyed by the `standard_name` attribute and some by additional (scalar) variables.

Because of this, we should tackle this variable by variable. The only variable that I have seen mentioned by name is `(fx, sftlf)`.
Thing is, to my best ability it seems that both [in the datarequest](http://clipc-services.ceda.ac.uk/dreq/u/bab742c8-e5dd-11e5-8482-ac72891c3257.html) and [in the table](http://clipc-services.ceda.ac.uk/dreq/u/bab742c8-e5dd-11e5-8482-ac72891c3257.html) I can not find an indication that a land type should be present. Therefore I think at the moment that models without it are correct and eg IPSL is wrong in including it.

To get a better picture, I'd like to know some of the problematic files. If you know some, please add them into the following list, organized by `(table, variable)` pairs.

* `(fx, sftlf)`

**Comment by zklaus** [comment]
Please also add example files if you have any! Eg as the `{dataset:...}` line.

**Comment by zklaus** [comment]
Digging a bit deeper it seems this is due to a change of the tables: Going from data spec version 01.00.29 to 01.00.30, on 11 March 2019, the `typeland` dimension vanished from `(fx, sftlf)` and the `typeli` dimension from `(fx, sftgif)`. The same commit also removes `typesea` from `(Ofx, sftof)`.

So these `fx`/`Ofx` files go back to #150, #158, and #159. Basically some models have prepared those files with older versions of the tables and in the newer versions these attributes are no longer present.

The `Lmon`/`Eyr` files that @bascrezee mentions are different though. But since they are clearly not `fx` files, I suggest we continue the discussion about those in #328 and focus here on the problems with the `fx` files and in general on those stemming from the change of tables, ok?

**Comment by valeriupredoi** [comment]
good digging @zklaus :wrench: @ValerioLembo has mentioned in #1272 that he's started seeing these quite a few - Valerio, care to add more files you've seen and are not mentioned here? :beer:

**Comment by ValerioLembo** [comment]
Hi there. 

I am following the discussion, but I am not replying because I am working in a branch I have opened, testing the fixes I have written for the few models I have spotted (namely, EC-Earth-Veg and GFDL-CM4). As soon as I am done with that, I will commit them and send the PR.

**Comment by zklaus** [comment]
To maximize the chance of implementing the right fix, I suggest to fully diagnose the problem first. For example I think that the fix suggested in  ESMValGroup/ESMValTool#1272 is wrong since it aligns the data with an outdated version of the data spec, whereas in the discussion in #158 (and #159) we agreed to align with the most recent version.

At first glance it seems indeed nonsensical to have, eg, a land type coordinate for sftlf, when there is patently no distinction between land types in that variable. Hence the removal seems to be a bugfix of the outdated version of the specs.

**Comment by zklaus** [comment]
The `height` stuff seems fine to me. I'd like to take a closer look at the type things first, so I suggest you separate the two and submit a pull request for the height stuff. Please add me as a reviewer since I already have some comments :smiling_imp:.

**Comment by ValerioLembo** [comment]
@zklaus there you go! #330

**Comment by zklaus** [comment]
As for the other type things: This seems to be because in ESMValCore we are at the moment also at the (slightly) outdated version of the data spec of 01.00.29 whereas the mismatching data files are by-and-large made with version 01.00.30.

So following the path set out in #158, #159, we should update to the latest version of the tables and then the errors will vanish. Mostly. Then we might need to write some fixes to align older model data like CESM2.

**Comment by bascrezee** [comment]
Thanks @zklaus for digging into this and explaining the different sort of tables, also in the other issues that you linked.

> As for the other type things: This seems to be because in ESMValCore we are at the moment also at the (slightly) outdated version of the data spec of 01.00.29 whereas the mismatching data files are by-and-large made with version 01.00.30.

If the version of the data spec is on the files (it is for the few models that I checked), we should check against the correct table, right? If there is a new table available, it should be added immediately to the tool. It would be so terribly bad if `CMIP beginners` (like me!) start fixing things that should not be fixed. If we build in a solid check against the right tables, this can not happen.

> Then we might need to write some fixes to align older model data like CESM2.
Why? If we make sure that the right tables are there, no fixes should be needed right? Or are there some technicalities that I overlook at the moment?

**Comment by zklaus** [comment]
This was discussed a little bit in https://github.com/ESMValGroup/ESMValCore/pull/158#issuecomment-510420934. That is a link to the specific comment since the entire discussion in that pr is a bit more long winded.

The problem with simply supporting different tables is twofold:

* It is technically and from a maintenance point of view rather demanding
* It leads to (potentially) inconsistent data: Should a diagnostic using `sftlf` deal with `typeland` or not? If we support all tables in the backend, every diagnostic has to be able to cope with all table versions as well.

If you factor in that most late changes to tables are probably done to address a specific problem, it seems more reasonable to make sure that all data that is presented to diagnostics follows the same version of the standard.

**Comment by zklaus** [comment]
To sum up:

* The fx stuff should be dealt with by addressing #159.
* The vegetation stuff was an individual model mistake that will be dealt with in #328. For this part, the tables have been stable for over 3 years (since version 01.00.05).

We can keep this issue open until the other two are closed, but otherwise I think this is settled.

**Comment by bascrezee** [comment]
> This was discussed a little bit in [#158 (comment)](https://github.com/ESMValGroup/ESMValCore/pull/158#issuecomment-510420934). That is a link to the specific comment since the entire discussion in that pr is a bit more long winded.
> 
> The problem with simply supporting different tables is twofold:
> 
> * It is technically and from a maintenance point of view rather demanding
> * It leads to (potentially) inconsistent data: Should a diagnostic using `sftlf` deal with `typeland` or not? If we support all tables in the backend, every diagnostic has to be able to cope with all table versions as well.

I totally missed the discussion in #158, thanks for pointing me towards this, since it is highly relevant. 

There are definitely some valid points made in favour of the approach decided in #158, however I am worried about backward compatibility. How do we guarantee that a diagnostic developed for CMIP5 data still works after we updated the core to align with the latest tables? Maybe CMIP7 will require a new attribute `tree_species` :deciduous_tree: , which is not available on any CMIP5 data. Or is a user supposed here to work with an older release of the ESMValTool ? 

Also, it would in fact imply that ESMValTool defines a new standard, a standard which 'continuously'  aligns with the latest CMOR standards. I think this might actually confuse diagnostic developers more than having to deal with different definitions. Given the `data_specs_version` attribute and the knowledge that one can find the definitions in the different github branches https://github.com/PCMDI/cmip6-cmor-tables a user can easily trace back if a dataset fully complies with the definitions.

**Comment by zklaus** [comment]
Let's take this discussion in #159.

**Comment by valeriupredoi** [comment]
so what do you propose @zklaus -> close this and move to #159 ? I also think we should actually think of seeking advice from people that work directly with the data *ie* ESGF folk, since we can tumble and twist for ages amongst themselves but not knowing in detail what the data publication/update plans are, we might end up doing a whole lot of soon-to-be obsolete work. I can get the ball rolling with the UK ESGF peeps if people think it's a good idea :beer:

There is also the side that @axel-lauer is on and I am starting to be more and more on that myself -> relax to max the CMOR checks

**Comment by zklaus** [comment]
Look. There is the issue of different versions of data spec. Modelling groups will not update their data unless there is an additional strong reason to do so, like a mistake in forcings that requires a rerun of the experiment or similar. Even then they will be reluctant to update their cmorization procedure.

So ESMValTool has to be able to deal with different data spec versions. Nothing ESGF or modelling folks will do anything about. And in principle this is also nothing that relaxed checking will help with.

How to deal with this and how to support it is exactly the topic of #159.

**Comment by valeriupredoi** [comment]
> Look. There is the issue of different versions of data spec. Modelling groups will not update their data unless there is an additional strong reason to do so, like a mistake in forcings that requires a rerun of the experiment or similar. Even then they will be reluctant to update their cmorization procedure

ESGF folk have tests that allow or reject the data for publication (eg prepare) - if we talk to them maybe we can find out what data specs and data request/CMOR versions they allow for in those tests; we can also tell them about all sorts of issues we found and they may not be aware of or willingly let them go through (it'd give us a feeling of how much leniency they actually allow for, one thing that's a mystery to my atm). Not saying we will find Mecca but hey, how much can such a discussion hurt? :beer:

**Comment by zklaus** [comment]
It hurts by wasting time. We know that they allow different data spec versions from looking at the data that passed. We know that their checks are poor from the amount of errors we find.

Of course, that doesn't mean we shouldn't talk. I think ESGF and the whole community has made big steps forward on the path from CMIP3 via CMIP5 to CMIP6.

But it is completely clear that there will be no change in CMIP6 that could mean that ESMValTool doesn't have to deal with data with different data specs versions, so we should address that need regardless of the conversation with the community.

**Comment by valeriupredoi** [comment]
ach K-man, you such a Prussian - attack the data now! :grin: Yes, I agree with all your points, it's just that it'd help us a heap more if we actually better understood what the heck those ESGF folk are up to. I can have a discussion with them anyways since we're working on the CMORization of UKESM1 data on Jasmin :beer:

**Comment by zklaus** [comment]
It's always good to understand better, so by all means go ahead. But do you agree that we can/should progress with the support of multiple data spec versions one way or another in the mean time?

**Comment by valeriupredoi** [comment]
oh yes absolutely! No question about that :beer:

**Comment by bascrezee** [comment]
> There is also the side that @axel-lauer is on and I am starting to be more and more on that myself -> relax to max the CMOR checks

Interesting point. I think the question to be asked here is whether or not the checking of CMOR standards is within the scope of the ESMValTool. See the input given in https://github.com/ESMValGroup/ESMValTool/issues/1027. I just keep pushing that thread :wink:  because I think it is so essential for the future development of ESMValTool !  

> So ESMValTool has to be able to deal with different data spec versions. Nothing ESGF or modelling folks will do anything about. And in principle this is also nothing that relaxed checking will help with.

I agree, ESMValTool has to be able to deal with different data spec versions, but I think the question still is at which level we want to deal with them. One can (a) deal with all of them at the preprocessor level, but one can also (b) relax checks and therefore leave inhomogeneities between datasets to diagnostic developers to deal with.

I am neutral :switzerland: :wink: with respect to the two options, but I do think it is good to give them some good thoughts and discuss them with the community before moving forward here too fast.

**Comment by zklaus** [comment]
I am reviewing this while checking all the open data issues. Since we have updated the tables in #363, is this still an issue? @ValerioLembo, could you please weigh in?

**Comment by ValerioLembo** [comment]
@zklaus I made a test with HadGEM3-GC31-LL and IPSL-CM6A-LR and they seem to work fine, so I would say that it should be fine now...

**Comment by zklaus** [comment]
Alright, closing then. Please reopen if issue resurfaces.
